### ISSUE TITLE
Loss D goes to zero when adding a new loss to generator

### ISSUE THREAD

**Issue by John1231983** [junyanz/pytorch-CycleGAN-and-pix2pix#626]
I have successfully run cycleGAN in my dataset. The D, G, cycle, idt loss are normal. 
However, when I add a new loss to the cyclegan. The discriminator loss easy goes down to 0, the results of the generator look terrible. It seems D easily judges the generated outputs when the D loss goes to 0. What is my problem? How should I fix it? I do not think D loss goes to zero is normal

**Comment by junyanz** [comment]
Maybe your newly added loss is too strong compared to existing CycleGAN losses. I=

**Comment by John1231983** [comment]
So, could you suggest any solutiom to solve the issue?

**Comment by junyanz** [comment]
Maybe reducing the weight for your new loss.

**Comment by John1231983** [comment]
@junyanz : Thanks. I have removed the new loss and the issue still maintain. I guess that cycleGAN has the issue when I use it on the new dataset. For traditional GAN, they said that loss D goes to zero mean the model is a [failure (tip 10](https://github.com/soumith/ganhacks). Do you know how to handle with the issue

**Comment by junyanz** [comment]
maybe making the discriminator weaker or the generator stronger.

**Comment by vrao9** [comment]
I'm also facing the same issue. I thought about increasing the number of times the generator gets updated when compared to the discriminator or decreasing the learning rate of the discriminator.
What are other ways in which generator can be made stronger / discriminator can be made weaker? @junyanz

**Comment by junyanz** [comment]
You can set different learning rates for G and D. See this [paper](https://arxiv.org/abs/1706.08500) for more details.

**Comment by vrao9** [comment]
Thank you

**Comment by John1231983** [comment]
@vrao9: do you find the good solution to solve the issue? I found that reduce number layer of D can solve it. I did not test your suggested solution: reduce lr, or increasing training G

**Comment by vrao9** [comment]
@John1231983 good to know that!
I tried 2 things:
1) reduce the learning rate: it did help in solving the problem
2) update the discriminator only when accuracy of prediction by the discriminator for fake image is less than 50%: there were less updates in the discriminator weights because of this new rule and discriminator loss did not go to zero.

**Comment by John1231983** [comment]
"accuracy of prediction by the discriminator for fake image is less than 50%: " how do you do it in pytorch? Thanks

**Comment by vrao9** [comment]
I used the keras implementation: 
https://github.com/eriklindernoren/Keras-GAN/tree/master/cyclegan
I don't know how it can be done in pytorch.

**Comment by John1231983** [comment]
@vrao9 : Thanks. Which line to do "accuracy of prediction by the discriminator for fake image is less than 50%: "?

**Comment by vrao9** [comment]
[Here](https://github.com/eriklindernoren/Keras-GAN/blob/44d3320e84ca00071de8a5c0fb4566d10486bb1d/cyclegan/cyclegan.py#L186), dB_loss_fake[1] contains the 'accuracy of prediction' for fake images.

**Comment by nehaleosharma** [comment]
How to print accuracy with the other details printed such as epoch and various losses.where to add metric=[accuracy]

**Comment by Legendsevl** [comment]
wow,tahks,but really what I want is a direct PDF file.




------------------ 原始邮件 ------------------
发件人: "nehaleosharma"<notifications@github.com>;
发送时间: 2019年7月30日(星期二) 下午4:54
收件人: "junyanz/pytorch-CycleGAN-and-pix2pix"<pytorch-CycleGAN-and-pix2pix@noreply.github.com>;
抄送: "Subscribed"<subscribed@noreply.github.com>;
主题: Re: [junyanz/pytorch-CycleGAN-and-pix2pix] Loss D goes to zero whenadding a new loss to generator (#626)




How to print accuracy with the other details printed such as epoch and various losses.where to add metric=[accuracy]

**Comment by JerryLeolfl** [comment]
> [Here](https://github.com/eriklindernoren/Keras-GAN/blob/44d3320e84ca00071de8a5c0fb4566d10486bb1d/cyclegan/cyclegan.py#L186), dB_loss_fake[1] contains the 'accuracy of prediction' for fake images.

Did you change the code in base keras '_train_on_batch_' to get the accuracy of prediction? Or just give it a parameter (_sample_weight=0.5_). I am not really understand how it operated.

**Comment by annihi1ation** [comment]
@junyanz 
hey professor?
Could you tell me if there is any suggection to weaker the discriminator instead of changing LR?

**Comment by annihi1ation** [comment]
@junyanz 
Or more concretely, I changed the BCE loss to MSE loss (according to the LSGAN). The model just failed, d loss is close to 0 while the g loss keeping increasing

**Comment by JunCS1** [comment]
> 2\. update the discriminator only when accuracy of prediction by the discriminator for fake image is less than 50%: there were less updates in the discriminator weights because of this new rule and discriminator loss did not go to zero.

Hi, may I ask if you used bce loss or mse loss? if you use mse loss, the loss won't be bounded between 0-1.
